Function streamlining

tadhg-ohiggins opened the pull request:
Work-in-progress on making functions easier to read and eventually reducing some of the issues that CodeClimate is complaining about.

lauraGgit commented:
@tadhg-ohiggins is this still do not merge?

tadhg-ohiggins commented:
@lauraGgit I think it's ready for review now.

lauraGgit commented on src/controllers/sudsconnection/populateFields.js:
```diff
@@ -151,18 +138,17 @@ function extractIntakeValue(intakeRequest, field, splitPath) {
  * @return {String|integer} - value of the field
  */
-function generateValue(field, intakeRequest, splitPath, person, fieldKey, autoPopulatedFields) {
+function generateValue(fieldKey, field, intakeRequest, { person, autoPopulatedKeys }) {
```
I would rather we take out the object if its just to appease the codeclimate report.

lauraGgit commented on src/controllers/sudsconnection/populateFields.js:
```diff
@@ -173,40 +159,32 @@ function generateValue(field, intakeRequest, splitPath, person, fieldKey, autoPo
+	Object.keys(fieldsByEndpoint).forEach((endpoint) => {
+		requestsTobeSent[endpoint] = {};
+		Object.keys(fieldsByEndpoint[endpoint]).forEach((fieldKey) => {
+			const field = fieldsByEndpoint[endpoint][fieldKey];
```
maybe its my screen but are the indents correct here?

tadhg-ohiggins replied:
I think I changed this to match what the linter wanted, but they look incorrect to me above too.

lauraGgit commented on src/controllers/sudsconnection/populateFields.js:
```diff
@@ -173,40 +159,32 @@ function generateValue(field, intakeRequest, splitPath, person, fieldKey, autoPo
+		Object.keys(fieldsByEndpoint[endpoint]).forEach((fieldKey) => {
+			const field = fieldsByEndpoint[endpoint][fieldKey];
+
+			const generatedValue = generateValue(fieldKey, field, intakeRequest, { person, autoPopulatedKeys });
```
is there are reason to not pass split path as we are using it in both functions?

lauraGgit commented on src/controllers/sudsconnection/post.js:
```diff
@@ -39,13 +39,13 @@ function isAppFromPerson(body){
 
 /**
  * Creates request for Basic API calls to create contact
```
let's update this to `SUDS`. I don't know how this escaped the find replace.

lauraGgit commented on src/controllers/sudsconnection/post.js:
```diff
@@ -68,13 +68,10 @@ function assignFieldsToEndpoints(fieldsToBasic){
 					fieldsAssignedToEndpoints[location] = {};
-					fieldsAssignedToEndpoints[location][key] = field[key];
 				}
+				fieldsAssignedToEndpoints[location][key] = field[key];
```
I'm a bit confused as to what is different here.

lauraGgit commented on src/controllers/sudsconnection/post.js:
```diff
@@ -89,7 +86,7 @@ function assignFieldsToEndpoints(fieldsToBasic){
 function prepareSudsPost(validationSchema, body, person){
 	const fieldsToPost = [];
-	db.getFieldsToStore(validationSchema, fieldsToPost, '', 'SUDS');
+	db.getFieldsToStore(validationSchema, fieldsToPost, '', 'SUDS'); // mutates fieldsToPost
```
I wonder if we should move this to the `util` controller, because it feels weird that's in `db`

lauraGgit commented on src/controllers/sudsconnection/post.js:
```diff
@@ -189,64 +170,38 @@ function createApplication(fieldsObj, contCN, apiCallLogObject, sudsToken){
-	{
-		apiCallLogObject.GET[contactGETOptions.logUri].response = res;
+function managePostContacts(apiCallLogObject, contactGETOptions, res, person, sudsToken, fieldsInSudsPostFormat) {
+	/* Set the apiCalLLogObject
```
You can add this comment in the docstring above as part of the description, just put a line break before the params: https://github.com/shri/JSDoc-Style-Guide

tadhg-ohiggins replied:
👍

lauraGgit commented on test/post-translation-test.js:
```diff
@@ -470,3 +475,221 @@ describe('Tests the populateValues function', function(){
+
+//*******************************************************************
+
+describe('tests the managePostContacts function.', function () {
```
love all these tests, maybe we want to put them in their own test file?

tadhg-ohiggins replied:
That's a good idea, there are a lot of them.

lauraGgit commented on test/post-translation-test.js:
```diff
@@ -470,3 +475,221 @@ describe('Tests the populateValues function', function(){
+		const testInfo = noncommercialFactory.create(testInfoUpdate);
+		const contactGetOptions = setContactGETOptions(testInfo.applicantInfo, person, token, apiObj);
+		const fieldsInSudsPostFormat = wrapSudsPrep(testInfo, noncommercialSchema, person);
+		const result = await managePostContacts(apiObj, contactGetOptions, response, person, token, fieldsInSudsPostFormat);
```
for most of the tests we've been using promise expectations rather than `await`. Maybe we want to change them to promises as thats how the functions will be called in the code?

tadhg-ohiggins replied:
I wasn't able to figure out how to get them to work as Promises, but I'm open to changing them to that form for consistency.

tadhg-ohiggins commented on src/controllers/sudsconnection/populateFields.js:
```diff
@@ -22,8 +22,7 @@ const logger = require('../utility.js').logger;
 function concat(input){
-	const output = input.join('');
-	return output;
+	return input.join('');
```
I think it's less confusing to keep it as a function with the same name as how we refer to it in `field.madeOf.function`.

lauraGgit replied:
sounds good.

tadhg-ohiggins commented on src/controllers/sudsconnection/populateFields.js:
```diff
@@ -106,23 +106,10 @@ function generateAutoPopulatedField(field, person, fieldMakeUp) {
-	
+function buildAutoPopulatedField (field, person, body) {
+	const getValue = (body, madeOfField) => madeOfField.fromIntake ? getFieldFromBody(madeOfField.field, body) : madeOfField.value;
+	const fieldMakeUp = field.madeOf.fields.map((madeOfField) => getValue(body, madeOfField)).filter(i => i);
```
Sure. I was uncertain about this one partly because I don't know the extent to which `.filter(i => i)` is used as idiomatic JS to remove falsy values.

tadhg-ohiggins commented:
@lauraGgit I addressed most of your comments, with the exceptions of:

- Moving `getFieldsToStore` out of `db`
- Switching the `managePostContacts` tests to only use Promises to handle asynchronous testing (because I couldn't seem to generate appropriate test failures with that, and am likely just missing something about how to set them up correctly).
